@Alyx/getFile.m

```diff
@@ -1,36 +1,58 @@
-function fullPath = getFile(obj, datasetURL)
-%GETFILE Recovers the full filepath of a file on the repository, given the datasetURL
-%	This function is SUPER inefficient because it has to load all
-%	filerecords from the database, and then search for the specific
-%	filerecord whose parent dataset matches the one supplied as input.
+function [fullPath, exists] = getFile(obj, eid, type)
+%GETFILE Returns the full filepath for a given eid.
+%	Returns the full path of associated data files given a dataset or file
+%	record URL or eid (experiment ID). Also returns a logical array 
+% indicating whether files exist for each element in the `eid` array.
 %
-%   datasetURL: URL of a dataset on Alyx
+% Inputs:
+%   `eid`: a string or char array containing the full URL(s) or eid(s) of 
+%   dataset(s) / file record(s) on Alyx.
+%   `type`: a string indicating type of eid(s) ('dataset' (default) or 
+%   'file')
 %
-% See also ALYX, GETDATA
+% Outputs:
+%   `fullPath`: a string array containing the full data file path(s) for
+%   each element in the `eid` array
+%   `exists`: a logical array indicating whether files exist for each eid
+%   in the `eid` array
+%
+% See also ALYX, GETDATA, GETSESSIONS
 %
 % Part of Alyx
-% TODO: Create endpoint so this function is no longer inefficient?
 % 2017 PZH created
+% 2019 MW Rewrote
 
+if nargin < 3; type = 'dataset'; end
 
-% Get all file records
-filerecords = obj.getData('files');
-
-% Extract the datasets which are parent to the filerecords
-datasets = cellfun(@(fr) fr.dataset, filerecords, 'uni', 0);
-
-% Find whichever filerecord has input datasetURL as its parent
-idx = contains(datasets, datasetURL);
-
-if any(idx)
-  fr = filerecords{idx};
-  relPath = fr.relative_path; % Get relative path of file
-  
-  repo = obj.getData('data-repository'); % Get absolute path of repository
-  fullPath = [repo{1}.path relPath]; % Recover the full path of the file on the repository
+% Convert array of strings to cell string
+if isstring(eid) && ~isscalar(eid)
+  eid = cellstr(eid);
 else
-  error('No filerecords with inputted datasetURL as its parent');
+  eid = ensureCell(eid);
 end
 
+% Validate URL
+assert(all(cellfun(@(str)ischar(str) && length(str) >= 36, eid)), 'Invalid eid')
+eid = mapToCell(@(str)str(end-35:end), eid);
```

Comment:
why is `mapToCell` necessary here?

Reply:
Because the input may be a callstr whose numel > 1.  The user may provide an eid or full url.  That's why the assert was >= and not ==.  I realized this just now as your last commit broke the method.  This is why webhooks are so useful!

Reply:
oh duh!